### PR TITLE
Configure compact source directories; use autoloader always

### PR DESCRIPTION
**Pull request by timriley**
In this PR, we update Hanami to depend on [dry-system 0.21.0](https://github.com/dry-rb/dry-system/releases/tag/v0.21.0) and use its new component dir namespaces to configure more compact component directories.

The file previously expected to live at e.g. `slices/main/lib/main/` can now reside in `slices/main/lib/`. The slice container will expect  the classes defined within all to be inside the `Main` constant namespace. These components will (still) be registered with top-level container keys. For example, `slices/main/lib/foo.rb` should define `Main::Foo`, and will be registered as `"foo"`.

In addition, we've added a new `component_dir_paths` setting that specifies additional component dirs to load from in the root of each slice's directory. By default, this is `%w[actions repositories views]`. Unlike `lib/`, these directories are expected to contain classes whose namespace matches their name, and will also be registered with that name in the container key. For example, `slices/main/actions/home/show.rb` should define `Main::Actions::Home::Show`, and will be registered as `"actions.home.show"`.

Along with this, we've changed some rules about code loading in the root of the application (i.e. outside any slice). Now, `lib/[application_namespace]/` will no longer be configured as a component dir for the app container. This is to encourage our users to put the majority of their logic in slices. If there are some components that should be shared between slices, they should go into a slice of their own, and that slice be imported into the other slices that require it. This will encourage higher cohesion in and between slices, and reduced coupling to the application itself. If you do need to register something in the app container, then you should set up a file in `config/boot/` to do so.

Although `lib/[application_namespace]/` is no longer a component dir, we're still configuring it with the autoloader, so classes defined inside that directory will be loadable without needing an explicit `require`.

The rest of the root `lib/` directory is added to `$LOAD_PATH`, so you should explicitly `require` and files in other parts of `lib/`.

A final change here is to make it so the autoloader is always on, and not something the user can disable. These new directory structures inside the slices are a significant part of how we want Hanami 2 to look and feel, and they're entirely dependent on the autoloader to function properly. Given how reliable and unobtrusive Zeitwerk is, we expect all users to be happy with it inside their Hanami applications. We strive to keep things configurable as much as possible, but at this point the autoloader is now an inextricable part of Hanami 2.

---

A next step after this PR will be to make the source directory and component directory configuration more flexible. We will be renaming `component_dir_paths` to `source_dir_paths` (or similar) and making it possible for you to configure any combination of autoloading as well as component auto-registration. At  this point it will then be possible to have a default `entities/` top-level directory within every slice, which we could not do within this PR, because entities should not be auto-registered into the container.

We're going to press ahead and release another alpha before this change, however, because the work in this PR (and in dry-system 0.21.0) is already a rather large change to Hanami's code loading, so we'd like to see that tested as early as possible.

### CODE REVIEW

File: lib/hanami/slice.rb [hunk 1]
```diff
@@ -115,12 +115,20 @@ def define_container
...
-              dir.default_namespace = namespace_path.tr(File::SEPARATOR, config.namespace_separator)
+              if application.configuration.autoloader
+                # If we're using an autoloader, expect all files in the slice's `lib/` to
+                # use the slice's const namespace
+                dir.namespaces.root const: namespace_path.tr(File::SEPARATOR, config.namespace_separator)
+              else
+                # Otherwise expect them to be inside `lib/[namespace]`, and make that take
+                # the root identifier namespace in the container
+                dir.namespaces.add namespace_path
+              end
```
Comment (timriley) [inline lib/hanami/slice.rb:126]: If we're not autoloading, then I think it's reasonable to expect users to have their slices set up so the file system path matches the nested constant (i.e. `lib/foo/blah.rb` defining `Foo::Blah`).

This is our escape hatch that still allows the system to behave properly even if the autoloader is disabled.

What I wonder, though, is whether we need to offer any further configurability than this? I feel like we probably don't, at least not right away, since it would result in a whole lot of complexity and would probably require us to build some higher-level configuration abstraction over a bunch of these dry-system settings.

I think what would probably be better is just making sure the idea we have of "bring your own container" works as intended. Right now it's not really something out tests exercise.

**Comment by timriley** [comment]
[Configure zeitwerk by default to support flattened directory structures](https://trello.com/c/7Jlep2QQ/126-configure-zeitwerk-by-default-to-support-flattened-directory-structures)

File: lib/hanami/slice.rb [hunk 2]
```diff
@@ -113,12 +113,69 @@ def define_container
...
+                #   should define Bar::Baz, and will be registered as "bar.baz"
+                component_dir.namespaces.add(namespace_path, key: nil)
+              end
             end
+          end
+
+          # When using an autoloader, add additional component dirs for each if the
+          # configured component dir paths (if they exist)
+          if application.configuration.autoloader
+            application.configuration.component_dir_paths.each do |slice_dir|
```
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/slice.rb:151]: > By "conflicting names" you mean file names, right? And so the idea of using $LOAD_PATH doesn't work because require "post" is ambiguous and may lead to unexpected results.
>
> Just double checking this with you, that we're talking about path name and not class name.
Because I expect these entities to have different namespaces: e.g. Admin::Entities::Post.

Yep, both of those statements are exactly right :)

Your idea about adding the slice root to the load path is an interesting one!

So to play this out fully, given these files:

```
slices/admin/entities/post.rb    # Admin::Entities::Post
slices/admin/lib/create_post.rb  # Admin::CreatePost

slices/main/entities/post.rb     # Main::Entities::Post
slices/main/lib/create_post.rb   # Main::CreatePost
```

We _would_ be able to properly require those entity files if we added the actual slices dir (i.e. `[app_root]/slices` itself, _not_ each individual slice dir) to the load path. Because in that case, `require "admin/entities/post" would properly maps onto the real filesystem path for the `slices/admin/entities/post.rb` file.

So while that works for the entities, there's still one last problem with this arrangement, in that there's still no way to require anything inside the `slices/[slice_name]/lib` directories, which is where we expect most of the per-slice business logic to be contained. For those files, we're stuck with the same problem I described initially, because:

1. There's no way for `require "admin/create_post"` to work (because `admin/create_post` does not map directly to any real filesystem path), and
2. For the same reasons as I originally described, we still can't add each slice's `lib/` directly onto `$LOAD_PATH`, because we'd have that same problem where it would then need `require "create_post"`, but that then becomes a require that will only work for whichever slice's `lib/` happens to be first on the `$LOAD_PATH`.

So this leads me back to those two options I shared before:

> 1. Make it so autoloading is always on, with no opt out
> 2. Keep the opt out in place, but accept and make it clear that if a user does opt out, they need to maintain a completely set of file hierarchies in their source directories

And my sense for the best option of these two is the same, too:

> Given that option (2) makes generators and documentation and many other things much more complicated, and given that I expect autoloading would be disabled only by the tiniest fraction of our users, I propose that we make it so that autoloading is always on.

What do you reckon?

File: lib/hanami/application.rb [hunk 3]
```diff
@@ -254,7 +254,8 @@ def configure_container(container)
 
           if root.join("lib").directory?
             config.component_dirs.add "lib" do |dir|
-              dir.default_namespace = application_name.to_s
+              # Register the files in `lib/[application_name]/` using top-level keys
+              dir.namespaces.add application_name.to_s, key: nil
```
[the remark this answers is not in the document]
Comment (timriley) [reply lib/hanami/application.rb]: Just for the record (for those following along), we decided not to go with this `app/` dir experiment, and I've taken it a step further by making it so that there is no default dir that auto-registers into the app container.

I've just pushed https://github.com/hanami/hanami/pull/1123/commits/4177d111f198cad79fd23dfdf40fa77f74f05c02, which:

- removes that `app/` experiment
- also removes all auto-registration of components into the app container
- sets up `lib/[app_namespace]/` with zeitwerk (so classes inside the app's own namespace can be autoloaded)
- adds `lib/` to the regular old load path

File: lib/hanami/slice.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jodosha) [reply lib/hanami/slice.rb]: @timriley I'm fine if they will be able to turn off autoloading, and need to manually do: `require "admin/lib/create_post"`.

We're talking about a tiny fraction of people who:

- Will disable autoloading
- Will put code outside of the canonical directories

If that require above looks wrong for them, they can still create an `admin/operations` directory, so the require will look better: `require "admin/operations/create_post"`.

---

To recap, I would:

- Give the ability to turn off autoloading
- Add `[root]/slices` to `$LOAD_PATH` to have nice manual requires from canonical directories. Example: `require "admin/entities/post"`.
- Ignore other ugly manual requires. Example: `require "admin/lib/create_post"`

**Review by jodosha — APPROVED** [review]

**Comment by jodosha** [comment]
For future reference, @timriley & I decided to not make possible to opt-out autoloading in Hanami 2.
